crates/sui-adapter/src/execution_engine.rs

```diff
@@ -155,7 +157,7 @@ fn execute_transaction<
         let execution_result = execution_loop::<Mode, _>(
             temporary_store,
             transaction_kind,
-            gas_object_id,
+            gas,
```

Comment:
It is not clear to me why this goes into the loop. 

Would it be possible to do something like
```
let gas_object_id = gas[0].0;
let gas_smash_changes = smash_gas(gas);
temporary_store.apply_object_changes(gas_smash_changes);
let result = charge_gas_for_object_read(temporary_store, &mut gas_status);
let mut result = result.and_then(|()| {
```
And then below
```
if execution_result.is_err() {
    temporary_store.apply_object_changes(gas_smash_changes);
    temporary_store.reset();
}
```

Long story short, programmable transactions do not work unless the gas is smashed before entering `execution_loop`. I also think doing it this will minimize the areas that need to change for execution?

Reply:
to be honest that is a bit the tragedy you have experienced as well.
That goes down only because Transfer uses it, one of the many special behavior of the gas flowing around the code base (waiting for ignition ;)).
I plan to readjust things when PT comes in and we clean up all the special behavior. I did not want to get intrusive.
If this creates a merge conflict I can adjust. 
Maybe I should study more how PT plans to use gas?

Reply:
I think I ended up doing what you mentioned here, take a look please